### DISCUSSION TITLE
Atomic Swaps: Advancing Decentralized Asset Exchange and Trust Minimization

### DISCUSSION THREAD

**Discussion by keyuno** [CounterpartyXCP/CIPs#100] [CIPs]
**Introduction:**
Atomic swaps enable direct peer-to-peer asset exchange between Bitcoin and Counterparty assets without the need for intermediaries or trusted third parties. They utilize Hash Time Locked Contracts to ensure secure and trustless transactions. By eliminating the vulnerabilities of dispensers and DEX limitations, atomic swaps offer a compelling alternative for users seeking secure, decentralized, and trustless asset trading.

**Challenges with Dispensers:**
Dispensers, which facilitate the exchange of assets, are prone to front-running, where malicious actors exploit transaction delays for personal gain. Additionally, some proposed solutions require trust in centralized verification processes, contradicting the principle of trust minimization.

**Counterparty DEX Limitations:**
Existing decentralized exchange (DEX) lacks support for asset ownership transfers and bundle sales. 

**Atomic Swaps:**
Atomic swaps present a superior solution to address these challenges. 

**Initial Setup**: Alice and Bob agree to exchange assets. Bob has Bitcoin (BTC), and Alice has a Counterparty asset (e.g., XCP). They decide on the exchange rate and the amount to be traded.

**Preimage Generation**: Alice generates a secret preimage that will be used to unlock the HTLCs. She keeps this preimage confidential and does not share it with Bob.

**Hash Locking**: Alice creates a hash of the secret preimage and provides it to Bob. Bob uses this hash to lock his HTLC. The hash acts as proof that the preimage exists without revealing the actual preimage.

**Hash Time Locked Contracts (HTLC)**: To facilitate the atomic swap, both Alice and Bob create Hash Time Locked Contracts (HTLCs). 

**Time Constraints**: The HTLCs have specific time constraints. Alice's HTLC has a longer time constraint compared to Bob's HTLC. This longer time constraint allows Bob enough time to claim his asset from Alice's locked HTLC after she reveals the preimage. This process adds an extra layer of security and prevents Alice from attempting to undercut the transaction before Bob can complete the swap.

**Verification**: Both parties verify each other's contracts to ensure that the agreed-upon conditions are accurately represented.

**Contract Funding**: Alice and Bob fund their respective HTLCs with the assets they want to trade. The assets are locked in the HTLCs but cannot be immediately spent.

**Claiming the Funds**: When Alice is ready to proceed with the atomic swap, she reveals her secret preimage by claiming the locked BTC from the HTLC Bob created. This action also allows Bob to claim the locked XCP from the HTLC that Alice created.

The use of HTLCs and time constraints ensures that the exchange is trustless and secure. If either party fails to fulfill their part of the deal, the funds in the HTLCs can be refunded to their respective owners, preventing potential losses.

**Code Review**:

Presented below is the code for your examination, tested with XCP, BTC, and the sweep function on testnet. 

https://github.com/keyuno/atomic-swap-xcp

Please note that this code is intended solely for testing purposes.

**Comment by Jpja** [comment]
This is really good!

I have some questions about the implementation;
* does it lock a specific utxo? If this is the case, it may not be double-spend proof bcs Counterparty is address-based, not utxo based.
* does it require a locking tx to confirm first, and then a second tx where Bob sends the BTC?
* the secret is revealed in the first tx, right? Otherwise it is impossible for Counterparty to escrow the assets.
* can you show specific tx's on testnet where an atomic swap has been made?

Please excuse me if the questions are stupid. I've read the proposal several times but can't totally wrap my head around it.

If this implementation won't work, a simpler atomic swap without a secret preimage is possible for sure.

**Reply by keyuno** [reply]
First and foremost, thank you for your questions. Engaging in these discussions is crucial as we progress towards achieving greater decentralization.

> does it lock a specific utxo? If this is the case, it may not be double-spend proof bcs Counterparty is address-based, not utxo based.

The locking script, is address-based. A Pay-to-Script-Hash (P2SH) address is generated from the locking script. This P2SH address is what will be used to receive and lock the funds. 

You can review the `lockingScript` [here](https://github.com/keyuno/atomic-swap-xcp/blob/main/bob.js).

To generate address:

```bash
node bob.js generate
```

The complete instructions are [here](https://github.com/keyuno/atomic-swap-xcp/blob/main/README.md).

> does it require a locking tx to confirm first, and then a second tx where Bob sends the BTC?

Yes, to begin the process, Alice initiates an XCP funding transaction that needs confirmation first. Afterwards, a second transaction takes place, with Bob initiating a BTC funding transaction.

Let’s go through the steps:

1. Alice initiates the process by sending XCP to a HTLC address that was created by Bob. This locked address is generated using a hash of the secret preimage that Alice provided to Bob earlier. At this point, the XCP is locked.

2. After the first transaction (XCP funding) is confirmed, Bob sends BTC to a separate HTLC address created by Alice. This address is also locked using the same hash of the preimage. 

3. Alice claims BTC by revealing the actual preimage (secret) in a new transaction. 

4. Once Alice has revealed the secret, Bob can now use it to claim XCP.

An example for the secret extraction script is provided [here](https://github.com/keyuno/atomic-swap-xcp/blob/main/secret.js).

> the secret is revealed in the first tx, right? Otherwise it is impossible for Counterparty to escrow the assets.

The secret is kept confidential until Alice decides to redeem BTC. 

Let's go through the steps:

1. Hash Locking: Alice creates a hash of the secret preimage and shares this hash with Bob. The hash can be used to verify the preimage without revealing it. Alice does not reveal the actual preimage at this point.

2. Alice funds an address with XCP and locks it with the condition that the XCP can only be released by Bob if Alice reveals the actual preimage that corresponds to the hash. 

3. Secret Reveal and Asset Claim: When Alice wants to claim BTC, she must include the actual preimage (secret) in the transaction. This secret is used to prove that she knows the preimage corresponding to the previously shared hash. By revealing the secret, Alice fulfills the condition of the script and allowed to claim BTC.

4. Unlocking XCP: After Alice reveals the secret, Bob can use the same secret to unlock the previously locked XCP by Alice.

It's important to note that a time lock is implemented to prevent Alice from taking advantage of her prior knowledge of the secret. In this process, Alice must wait longer than Bob to make a refund transaction. 

In case the deal fails, and Alice doesn't reveal the secret to claim the BTC on time, Bob can initiate a refund transaction from the address he generated. 

The extended waiting period ensures that she cannot claim the BTC without providing Bob with enough time to claim the locked XCP.

> can you show specific tx's on testnet where an atomic swap has been made?

This is the generated P2SH address using the locking script created by Alice:

2N7jK12XQk3Dh2hHjNsnUgDkYWbuGTX9tun

Alice funds the Address with the agreed amount of XCP plus some BTC for transaction fees:

XCP funding [transaction](https://testnet.xchain.io/tx/c5ede798e175e3fb582d4fa3e4a7a11da0781a2b04fb3b8b66e0396ec87e785f).

BTC [transaction ](https://blockstream.info/testnet/address/2N7jK12XQk3Dh2hHjNsnUgDkYWbuGTX9tun) for fees.

This is the generated P2SH address using the locking script created by Bob:

2N7wTQhAYygpoHexpHCsXNrKdMX5T843FWy

Bob funds the address with agreed amount of BTC that includes the transaction fees.

BTC [transaction](https://blockstream.info/testnet/tx/d9ce605ecdc455802f11ea0d689b8247bc81e8f1b5c4c9a7af0bb02621146cc7).

### The Swap:

This [transaction](https://blockstream.info/testnet/tx/20e10ada695c3336d6c810632f34c7304a862e1849b22d96504c5a6aff15e493) is known as the redeem transaction initiated by Alice, in which she claims the BTC to her own address and reveals the secret in the process.

This [transaction](https://testnet.xchain.io/tx/39fdb736bdec453dcd1ce0ad4ed7ff9fc9728be4fdb3305e1e924e3ca68f98f1) is the redeem transaction conducted by Bob, where he claims the XCP to his own address using the secret that was revealed by Alice.

The atomic swap was a success.

> a simpler atomic swap without a secret preimage is possible for sure.

If you could provide a simpler implementation of an atomic swap, whether in a conceptual framework or actual code, I would greatly appreciate it.

**Reply by Jpja** [reply]
> If you could provide a simpler implementation of an atomic swap, whether in a conceptual framework or actual code, I would greatly appreciate it.

From the simplest to more advanced implementations of atomic swap (where Alice sells asset and Bob buys with bitcoin):

A) Similar to dispenser, except Alice specifies that only Bob can make the payment. And it must be done before block N. This will be useful after off-chain negotiations, for example to settle an auction.

B) The same but Alice sells several different types of tokens.

C) The same but also asset ownerships can be sold.

D) Alice can specify that Bob shall pay a list of assets in addition to sending BTC. While a simple bitcoin send will suffice for A-C, with D it becomes necessary for Bob to include an op_return similar to BTC-PAY. Otherwise Alice could steal Bob's assets by opening a swap in anticipation of a regular bitcoin transfer.

Only two transactions are necessary. Alice opens swap and the protocol escrows Alice's assets. Bob makes payment after Alice's tx confirms.

E) Your suggestion surely is more elegant. Did I understand correctly, this is done with Bitcoin scripting only – no update needed for Counterparty?

Suggestions A-D will require a new transaction type for Counterparty.

**Reply by keyuno** [reply]
> Did I understand correctly, this is done with Bitcoin scripting only – no update needed for Counterparty?

Achieved using Bitcoin scripting. No additional updates needed. Successfully tested on testnet.

>Suggestions A-D will require a new transaction type for Counterparty.

Currently, I'm more inclined to investigate the existing possibilities. Nonetheless, thank you for your suggestions.

**Comment by davestaxcp** [comment]
First off, this process looks very interesting and I am intrigued, though, I have a question:

Could this process be used to exchange multiple assets for both parties? 

For example Alice wants to trade 1.0 XCP, and an XCP asset she minted (ALICESTOKEN) to Bob.

Bob in exchange will give in return an XCP asset he minted (BOBSTOKEN) and BTC (0.0001 total) as well.

Would this multi-asset exchange be possible with Atomic Swaps?

**Reply by keyuno** [reply]
> Could this process be used to exchange multiple assets for both parties?

You can use `create_send` to construct a transaction with multiple assets. 

"_To send multiple assets/destinations simultaneously you can pass an array of parameters to the destination, asset and quantity parameters. If one of these parameters is an array then the other must be an array of equal length. Each entry corresponds to the same entry index on the other arrays._" [counterparty-server API
](https://github.com/CounterpartyXCP/Documentation/blob/master/Developers/API.md)

To send both Bitcoin and assets, you'll need to conduct separate transactions: one for assets and another for Bitcoin (it's [possible ](https://blockstream.info/testnet/tx/65770f1dbac699f9b089b821f351fbd3fa1d7bf3ac8e5870698c08b100868ef8)to send a **single** asset and Bitcoin in one go by adjusting the dust amount carefully).



If you wish to send and transfer ownership of Counterparty assets in a single transaction, you should consider using `create_sweep`, but keep in mind that it requires a higher cost due to the 0.5 XCP fee ([sweep demo](https://github.com/keyuno/atomic-swap-xcp/blob/main/sweep.js)).

**Comment by Jpja** [comment]
@keyuno I suggest making Atomic Swaps into a CIP.

Can be either an informative CIP - only need to format the text into the CIP format.

Or if you like, a CIP for adding Atomic Swaps to the reference client. In this case need bounties for dev and peer review.

**Reply by keyuno** [reply]
Thank you for your suggestion. I think additional testing is necessary. I am presently working on expanding atomic swaps functionality and will soon make it available for further experimentation. After this phase, we can proceed with standardization.
